{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Feb  5 22:42:48 PST 2020\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.patches as mpatches\n",
    "import scanpy as scanp\n",
    "from scipy.stats import ks_2samp, ttest_ind\n",
    "from scipy.sparse import csr_matrix\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "from sklearn.manifold import TSNE\n",
    "from umap import UMAP\n",
    "from sklearn.cluster import KMeans\n",
    "from scipy import stats\n",
    "from scipy.optimize import curve_fit\n",
    "import ast\n",
    "from scipy.stats import sem\n",
    "from sklearn.metrics import adjusted_rand_score\n",
    "from scipy.stats import ks_2samp, ttest_ind\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "fsize=20\n",
    "\n",
    "plt.rcParams.update({'font.size': fsize})\n",
    "%config InlineBackend.figure_format = 'retina'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "gene = anndata.read_h5ad(\"cell_ranger_annotation/gene.h5ad\")\n",
    "isoform = anndata.read_h5ad(\"cell_ranger_annotation/isoform.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# New t2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = [\"transcript_id\", \"gene_id\", \"gene_name\", \"transcript_name\", \"chromosome\", \"start\", \"stop\", \"strand\"]\n",
    "t2g = pd.read_csv(\"/home/sina/projects/mop/reference/cellranger_index/refdata-cellranger-mm10-3.0.0/kb_out/t2g.txt\", sep=\"\\t\", header=None, names=cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def swap_start(start, stop, strand):\n",
    "    new_start = start\n",
    "    new_stop = stop\n",
    "    if strand==\"-\":\n",
    "        new_start = stop\n",
    "        new_stop = start\n",
    "    return new_start, new_stop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = t2g.apply(lambda x: swap_start(x[\"start\"], x[\"stop\"], x[\"strand\"]), axis=1).tolist()\n",
    "t2g[[\"new_start\", \"new_stop\"]] = pd.DataFrame(tmp, columns=[\"new_start\", \"new_stop\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Group isoforms by (1) Gene and then (2) Transcripton start sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gb = t2g.groupby([\"chromosome\", \"gene_id\", \"new_start\"])[\"transcript_id\"].apply(list)\n",
    "gb = gb.reset_index()\n",
    "gb = gb.rename(columns={\"transcript_id\":\"tss_same\"})\n",
    "gb[\"n_tss_same\"] = gb[\"tss_same\"].apply(lambda x: len(x))\n",
    "num_iso_per_gene = t2g.groupby(\"gene_id\")[\"transcript_id\"].nunique()\n",
    "gb[\"n_iso\"] = gb[\"gene_id\"].map(num_iso_per_gene)\n",
    "txn_list = gb[\"tss_same\"].explode().values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>new_start</th>\n",
       "      <th>tss_same</th>\n",
       "      <th>n_tss_same</th>\n",
       "      <th>n_iso</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000000544.14</td>\n",
       "      <td>166130238</td>\n",
       "      <td>[ENSMUST00000060833.13]</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000000544.14</td>\n",
       "      <td>166130340</td>\n",
       "      <td>[ENSMUST00000166159.1]</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000000544.14</td>\n",
       "      <td>166130417</td>\n",
       "      <td>[ENSMUST00000027847.6]</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000000544.14</td>\n",
       "      <td>166130467</td>\n",
       "      <td>[ENSMUST00000166860.1]</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000000817.10</td>\n",
       "      <td>161788358</td>\n",
       "      <td>[ENSMUST00000193648.1]</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102881</th>\n",
       "      <td>Y</td>\n",
       "      <td>ENSMUSG00000104500.5</td>\n",
       "      <td>24276742</td>\n",
       "      <td>[ENSMUST00000194949.5]</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102882</th>\n",
       "      <td>Y</td>\n",
       "      <td>ENSMUSG00000104505.5</td>\n",
       "      <td>12516355</td>\n",
       "      <td>[ENSMUST00000191901.5, ENSMUST00000193731.1]</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102883</th>\n",
       "      <td>Y</td>\n",
       "      <td>ENSMUSG00000104505.5</td>\n",
       "      <td>12516809</td>\n",
       "      <td>[ENSMUST00000191742.5]</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102884</th>\n",
       "      <td>Y</td>\n",
       "      <td>ENSMUSG00000104506.5</td>\n",
       "      <td>19983060</td>\n",
       "      <td>[ENSMUST00000193523.5, ENSMUST00000194757.1]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102885</th>\n",
       "      <td>Y</td>\n",
       "      <td>ENSMUSG00000104522.5</td>\n",
       "      <td>13338145</td>\n",
       "      <td>[ENSMUST00000194145.5, ENSMUST00000193656.1]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102886 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome                gene_id  new_start  \\\n",
       "0               1  ENSMUSG00000000544.14  166130238   \n",
       "1               1  ENSMUSG00000000544.14  166130340   \n",
       "2               1  ENSMUSG00000000544.14  166130417   \n",
       "3               1  ENSMUSG00000000544.14  166130467   \n",
       "4               1  ENSMUSG00000000817.10  161788358   \n",
       "...           ...                    ...        ...   \n",
       "102881          Y   ENSMUSG00000104500.5   24276742   \n",
       "102882          Y   ENSMUSG00000104505.5   12516355   \n",
       "102883          Y   ENSMUSG00000104505.5   12516809   \n",
       "102884          Y   ENSMUSG00000104506.5   19983060   \n",
       "102885          Y   ENSMUSG00000104522.5   13338145   \n",
       "\n",
       "                                            tss_same  n_tss_same  n_iso  \n",
       "0                            [ENSMUST00000060833.13]           1      4  \n",
       "1                             [ENSMUST00000166159.1]           1      4  \n",
       "2                             [ENSMUST00000027847.6]           1      4  \n",
       "3                             [ENSMUST00000166860.1]           1      4  \n",
       "4                             [ENSMUST00000193648.1]           1      2  \n",
       "...                                              ...         ...    ...  \n",
       "102881                        [ENSMUST00000194949.5]           1      3  \n",
       "102882  [ENSMUST00000191901.5, ENSMUST00000193731.1]           2      3  \n",
       "102883                        [ENSMUST00000191742.5]           1      3  \n",
       "102884  [ENSMUST00000193523.5, ENSMUST00000194757.1]           2      2  \n",
       "102885  [ENSMUST00000194145.5, ENSMUST00000193656.1]           2      2  \n",
       "\n",
       "[102886 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gb = gb.loc[gb.index.repeat(gb.n_TSS_same)].reset_index(drop=True)\n",
    "#gb[\"transcript_id\"] = txn_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now GB has TSS_same which is the list of transcripts in the covering gene that have the same start site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "gb[\"tss_same\"] = gb.tss_same.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gb.drop_duplicates(subset=[\"chromosome\", \"gene_id\", \"tss_same\"]).shape\n",
    "gb[\"tss_class\"] = np.arange(gb.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Transcripts are grouped by Chromosome -> Gene -> TSS same\n",
    "\n",
    "Want to make a matrix which is cells x tss_same\n",
    "\n",
    "Then perform t test between clusters, this tells you which TSS same mark the cluster\n",
    "\n",
    "but want to compare TSS same with just isoforms only... am missing something, ah yes, sleep!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>new_start</th>\n",
       "      <th>tss_same</th>\n",
       "      <th>n_tss_same</th>\n",
       "      <th>n_iso</th>\n",
       "      <th>tss_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000000544.14</td>\n",
       "      <td>166130238</td>\n",
       "      <td>['ENSMUST00000060833.13']</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000000544.14</td>\n",
       "      <td>166130340</td>\n",
       "      <td>['ENSMUST00000166159.1']</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000000544.14</td>\n",
       "      <td>166130417</td>\n",
       "      <td>['ENSMUST00000027847.6']</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000000544.14</td>\n",
       "      <td>166130467</td>\n",
       "      <td>['ENSMUST00000166860.1']</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000000817.10</td>\n",
       "      <td>161788358</td>\n",
       "      <td>['ENSMUST00000193648.1']</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome                gene_id  new_start                   tss_same  \\\n",
       "0          1  ENSMUSG00000000544.14  166130238  ['ENSMUST00000060833.13']   \n",
       "1          1  ENSMUSG00000000544.14  166130340   ['ENSMUST00000166159.1']   \n",
       "2          1  ENSMUSG00000000544.14  166130417   ['ENSMUST00000027847.6']   \n",
       "3          1  ENSMUSG00000000544.14  166130467   ['ENSMUST00000166860.1']   \n",
       "4          1  ENSMUSG00000000817.10  161788358   ['ENSMUST00000193648.1']   \n",
       "\n",
       "   n_tss_same  n_iso  tss_class  \n",
       "0           1      4          0  \n",
       "1           1      4          1  \n",
       "2           1      4          2  \n",
       "3           1      4          3  \n",
       "4           1      2          4  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gb.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Now create one t2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Conditions\n",
    "# 1. n_iso per gene > 2\n",
    "# 2. 2< n_TSS_same < n_iso\n",
    "\n",
    "mask = np.logical_and(gb[\"n_tss_same\"].values >2, np.logical_and(gb[\"n_iso\"]>2, gb[\"n_tss_same\"].values < gb[\"n_iso\"]))\n",
    "\n",
    "f = gb[mask]\n",
    "f[\"tss_same\"] = f[\"tss_same\"].apply(lambda x: ast.literal_eval(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>new_start</th>\n",
       "      <th>tss_same</th>\n",
       "      <th>n_tss_same</th>\n",
       "      <th>n_iso</th>\n",
       "      <th>tss_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445271</td>\n",
       "      <td>[ENSMUST00000125304.7, ENSMUST00000115011.7, E...</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000003458.12</td>\n",
       "      <td>172082795</td>\n",
       "      <td>[ENSMUST00000003550.10, ENSMUST00000140643.7, ...</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000005763.15</td>\n",
       "      <td>165788681</td>\n",
       "      <td>[ENSMUST00000161971.7, ENSMUST00000187313.6, E...</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000005886.14</td>\n",
       "      <td>13374083</td>\n",
       "      <td>[ENSMUST00000068304.12, ENSMUST00000006037.12,...</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000006014.16</td>\n",
       "      <td>150465598</td>\n",
       "      <td>[ENSMUST00000006171.15, ENSMUST00000164600.7, ...</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101094</th>\n",
       "      <td>X</td>\n",
       "      <td>ENSMUSG00000067768.12</td>\n",
       "      <td>73214349</td>\n",
       "      <td>[ENSMUST00000069077.9, ENSMUST00000124904.7, E...</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>101094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101150</th>\n",
       "      <td>X</td>\n",
       "      <td>ENSMUSG00000071753.12</td>\n",
       "      <td>61116378</td>\n",
       "      <td>[ENSMUST00000232283.1, ENSMUST00000144853.8, E...</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>101150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101151</th>\n",
       "      <td>X</td>\n",
       "      <td>ENSMUSG00000071753.12</td>\n",
       "      <td>61116386</td>\n",
       "      <td>[ENSMUST00000137153.7, ENSMUST00000140023.7, E...</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>101151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101323</th>\n",
       "      <td>X</td>\n",
       "      <td>ENSMUSG00000079481.11</td>\n",
       "      <td>101849385</td>\n",
       "      <td>[ENSMUST00000125115.1, ENSMUST00000124279.5, E...</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>101323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101543</th>\n",
       "      <td>X</td>\n",
       "      <td>ENSMUSG00000087403.9</td>\n",
       "      <td>152327495</td>\n",
       "      <td>[ENSMUST00000195280.2, ENSMUST00000148326.7, E...</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>101543</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1035 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome                gene_id  new_start  \\\n",
       "14              1  ENSMUSG00000001143.13   36445271   \n",
       "46              1  ENSMUSG00000003458.12  172082795   \n",
       "151             1  ENSMUSG00000005763.15  165788681   \n",
       "164             1  ENSMUSG00000005886.14   13374083   \n",
       "191             1  ENSMUSG00000006014.16  150465598   \n",
       "...           ...                    ...        ...   \n",
       "101094          X  ENSMUSG00000067768.12   73214349   \n",
       "101150          X  ENSMUSG00000071753.12   61116378   \n",
       "101151          X  ENSMUSG00000071753.12   61116386   \n",
       "101323          X  ENSMUSG00000079481.11  101849385   \n",
       "101543          X   ENSMUSG00000087403.9  152327495   \n",
       "\n",
       "                                                 tss_same  n_tss_same  n_iso  \\\n",
       "14      [ENSMUST00000125304.7, ENSMUST00000115011.7, E...           4      8   \n",
       "46      [ENSMUST00000003550.10, ENSMUST00000140643.7, ...           4      6   \n",
       "151     [ENSMUST00000161971.7, ENSMUST00000187313.6, E...           3      9   \n",
       "164     [ENSMUST00000068304.12, ENSMUST00000006037.12,...           3     11   \n",
       "191     [ENSMUST00000006171.15, ENSMUST00000164600.7, ...           4      8   \n",
       "...                                                   ...         ...    ...   \n",
       "101094  [ENSMUST00000069077.9, ENSMUST00000124904.7, E...           3      8   \n",
       "101150  [ENSMUST00000232283.1, ENSMUST00000144853.8, E...           3     11   \n",
       "101151  [ENSMUST00000137153.7, ENSMUST00000140023.7, E...           5     11   \n",
       "101323  [ENSMUST00000125115.1, ENSMUST00000124279.5, E...           3      7   \n",
       "101543  [ENSMUST00000195280.2, ENSMUST00000148326.7, E...           3      7   \n",
       "\n",
       "        tss_class  \n",
       "14             14  \n",
       "46             46  \n",
       "151           151  \n",
       "164           164  \n",
       "191           191  \n",
       "...           ...  \n",
       "101094     101094  \n",
       "101150     101150  \n",
       "101151     101151  \n",
       "101323     101323  \n",
       "101543     101543  \n",
       "\n",
       "[1035 rows x 7 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Transcriptional Regulation (sum isoforms with same TSS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = f.explode(\"tss_same\")\n",
    "tmp[\"tss_same\"] = tmp.tss_same.apply(lambda x: x.split(\".\")[0])\n",
    "tmp[\"gene_id\"] = tmp.gene_id.apply(lambda x: x.split(\".\")[0])\n",
    "tmp.index = tmp.tss_same.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>new_start</th>\n",
       "      <th>tss_same</th>\n",
       "      <th>n_tss_same</th>\n",
       "      <th>n_iso</th>\n",
       "      <th>tss_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000001143</td>\n",
       "      <td>36445271</td>\n",
       "      <td>ENSMUST00000125304</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000001143</td>\n",
       "      <td>36445271</td>\n",
       "      <td>ENSMUST00000115011</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000001143</td>\n",
       "      <td>36445271</td>\n",
       "      <td>ENSMUST00000001171</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000001143</td>\n",
       "      <td>36445271</td>\n",
       "      <td>ENSMUST00000123583</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000003458</td>\n",
       "      <td>172082795</td>\n",
       "      <td>ENSMUST00000003550</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101323</th>\n",
       "      <td>X</td>\n",
       "      <td>ENSMUSG00000079481</td>\n",
       "      <td>101849385</td>\n",
       "      <td>ENSMUST00000124279</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>101323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101323</th>\n",
       "      <td>X</td>\n",
       "      <td>ENSMUSG00000079481</td>\n",
       "      <td>101849385</td>\n",
       "      <td>ENSMUST00000101339</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>101323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101543</th>\n",
       "      <td>X</td>\n",
       "      <td>ENSMUSG00000087403</td>\n",
       "      <td>152327495</td>\n",
       "      <td>ENSMUST00000195280</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>101543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101543</th>\n",
       "      <td>X</td>\n",
       "      <td>ENSMUSG00000087403</td>\n",
       "      <td>152327495</td>\n",
       "      <td>ENSMUST00000148326</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>101543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101543</th>\n",
       "      <td>X</td>\n",
       "      <td>ENSMUSG00000087403</td>\n",
       "      <td>152327495</td>\n",
       "      <td>ENSMUST00000149098</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>101543</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3869 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome             gene_id  new_start            tss_same  \\\n",
       "14              1  ENSMUSG00000001143   36445271  ENSMUST00000125304   \n",
       "14              1  ENSMUSG00000001143   36445271  ENSMUST00000115011   \n",
       "14              1  ENSMUSG00000001143   36445271  ENSMUST00000001171   \n",
       "14              1  ENSMUSG00000001143   36445271  ENSMUST00000123583   \n",
       "46              1  ENSMUSG00000003458  172082795  ENSMUST00000003550   \n",
       "...           ...                 ...        ...                 ...   \n",
       "101323          X  ENSMUSG00000079481  101849385  ENSMUST00000124279   \n",
       "101323          X  ENSMUSG00000079481  101849385  ENSMUST00000101339   \n",
       "101543          X  ENSMUSG00000087403  152327495  ENSMUST00000195280   \n",
       "101543          X  ENSMUSG00000087403  152327495  ENSMUST00000148326   \n",
       "101543          X  ENSMUSG00000087403  152327495  ENSMUST00000149098   \n",
       "\n",
       "        n_tss_same  n_iso  tss_class  \n",
       "14               4      8         14  \n",
       "14               4      8         14  \n",
       "14               4      8         14  \n",
       "14               4      8         14  \n",
       "46               4      6         46  \n",
       "...            ...    ...        ...  \n",
       "101323           3      7     101323  \n",
       "101323           3      7     101323  \n",
       "101543           3      7     101543  \n",
       "101543           3      7     101543  \n",
       "101543           3      7     101543  \n",
       "\n",
       "[3869 rows x 7 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.var` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "trans = isoform[:, isoform.var.transcript_id.isin(tmp[\"tss_same\"])]\n",
    "trans.var[\"tss_class\"] = trans.var.transcript_id.map(tmp[\"tss_class\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sum_tss_same(mtx, components, features):\n",
    "    \n",
    "    unique = np.unique(features)\n",
    "    nuniq = unique.shape[0]\n",
    "    ncomp = components.shape[0]\n",
    "    \n",
    "    s_mtx = np.zeros((ncomp, nuniq))\n",
    "    \n",
    "    for uidx, u in enumerate(unique):\n",
    "        mask = features == u\n",
    "        s_mtx[:,uidx] = np.asarray(mtx[:,mask].sum(axis=1)).reshape(-1)\n",
    "    \n",
    "    return s_mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sum the isoforms with the same start site\n",
    "features = trans.var.tss_class.values\n",
    "components = trans.obs.cell_id.values\n",
    "mtx = trans.layers[\"log1p\"].todense()\n",
    "# sum the raw counts then normalize and log?\n",
    "\n",
    "mtx = sum_tss_same(mtx, components, features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = mtx\n",
    "obs = isoform.obs\n",
    "var = trans.var.drop_duplicates([\"gene_id\", \"tss_class\"]).iloc[\n",
    "    np.argsort(trans.var.drop_duplicates([\"gene_id\", \"tss_class\"]).tss_class)]\n",
    "stss = anndata.AnnData(X=X, obs=obs, var=var)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STSS is an anndata where the TSS transcripts are grouped by start site and summed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "\n",
    "logger = logging.getLogger()\n",
    "logger.setLevel(logging.DEBUG)\n",
    "logging.basicConfig(format='%(asctime)s - %(message)s', datefmt='%d-%b-%y %H:%M:%S')\n",
    "\n",
    "def normalize_rows_L1(mat):\n",
    "    L1 = mat.sum(axis=1)\n",
    "    norm = mat/L1[:,None]\n",
    "    return norm\n",
    "\n",
    "def split_by_target(mat, targets, target, axis=0):\n",
    "    \"\"\"\n",
    "    Split the rows of mat by the proper assignment\n",
    "    \n",
    "    mat = ndarray\n",
    "    targets, length is equal to number of components (axis=0) or features (axis=1)\n",
    "    target is a singular element from unique(assignments/features) \n",
    "    \"\"\"\n",
    "    if axis==0 and len(targets) != mat.shape[axis]: return -1\n",
    "    if axis==1 and len(targets) != mat.shape[axis]: return -1\n",
    "        \n",
    "    mask = targets == target\n",
    "    \n",
    "    if axis==0:\n",
    "        t_mat = mat[mask]  # target matrix\n",
    "        c_mat = mat[~mask] # complement matrix\n",
    "    elif axis==1:\n",
    "        t_mat = mat[:, mask]  # target matrix\n",
    "        c_mat = mat[:, ~mask] # complement matrix\n",
    "    \n",
    "    return (t_mat, c_mat)\n",
    "\n",
    "def deSS(mtx, components, features, assignments, **kwargs):\n",
    "    nan_cutoff = kwargs.get(\"nan_cutoff\", 0)\n",
    "    corr_method = kwargs.get(\"corr_method\", \"bonferroni\")\n",
    "    \n",
    "    unique = np.unique(assignments)\n",
    "    unique_features = np.unique(features)\n",
    "    \n",
    "    nuniq = unique.shape[0]\n",
    "    nuniq_feat = unique_features.shape[0]\n",
    "    nfeat = features.shape[0]\n",
    "    \n",
    "    pval_bool = np.empty((nuniq, nuniq_feat)).astype(bool)\n",
    "    pval_bool.fill(False)\n",
    "    \n",
    "    for uidx, u in enumerate(unique):\n",
    "        logging.debug(\"{} of {} assignments: {}\".format(uidx+1, nuniq, u))\n",
    "        t_mat,c_mat = split_by_target(mtx, assignments, u)\n",
    "\n",
    "        # now iterate through the TSS classes\n",
    "        for tidx, t in enumerate(unique_features):\n",
    "            mask = features == t\n",
    "            x = t_mat[:,mask]\n",
    "            y = c_mat[:,mask]\n",
    "            \n",
    "            x = normalize_rows_L1(np.asarray(x))\n",
    "            y = normalize_rows_L1(np.asarray(y))\n",
    "            \n",
    "            all_cols_pass_nan =( (np.isnan(x).sum(axis=0)/x.shape[0])>nan_cutoff).sum() == x.shape[1] \n",
    "            \n",
    "            # perform sidak correction and \n",
    "            if all_cols_pass_nan:\n",
    "                test = ttest_ind(x,y, nan_policy=\"omit\")\n",
    "                if (test.pvalue<0.01).sum()==test.pvalue.shape[0]:\n",
    "                    pval_bool[uidx, tidx] = True\n",
    "    \n",
    "    return pval_bool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Once we map the transcripts to the isoform matrix, some of them were filtered? so i remove the ones with only 1 count\n",
    "mask = same.var.TSS_class.map(same.var.groupby(\"TSS_class\")[\"transcript_id\"].count())>1\n",
    "same = same[:,mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 514,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp1 = same[same.obs.subclass_label == \"Vip\"]\n",
    "tmp2 = same[same.obs.subclass_label != \"Vip\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 480,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs Ã— n_vars = 4724 Ã— 124 \n",
       "    obs: 'cluster_id', 'cluster_label', 'subclass_label', 'class_label', 'cluster_color', 'size', 'cell_id', 'cell_TPM', 'leiden', 'louvain'\n",
       "    var: 'transcript_id', 'gene_id', 'gene_name', 'transcript_name', 'gene_TPM', 'TSS_class'\n",
       "    uns: 'leiden', 'louvain', 'neighbors'\n",
       "    obsm: 'X_pca', 'X_tsne'\n",
       "    layers: 'X', 'log1p'"
      ]
     },
     "execution_count": 480,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 538,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "19-Feb-20 22:07:07 - 1 of 4 assignments: L5 IT\n",
      "19-Feb-20 22:07:07 - 2 of 4 assignments: Lamp5\n",
      "19-Feb-20 22:07:07 - 3 of 4 assignments: Sst\n",
      "19-Feb-20 22:07:07 - 4 of 4 assignments: Vip\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 47 ms, sys: 7.48 ms, total: 54.5 ms\n",
      "Wall time: 50.2 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "nc = 5\n",
    "nf = 10\n",
    "\n",
    "mat         = same.layers[\"log1p\"].todense()[:nc,:nf]\n",
    "components  = same.obs.cell_id.values[:nc]\n",
    "features    = same.var.TSS_class.values[:nf]\n",
    "assignments = same.obs.subclass_label.values[:nc]\n",
    "\n",
    "z = deSS(mat, components, features, assignments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 552,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "19-Feb-20 22:08:36 - 1 of 18 assignments: Astro\n",
      "19-Feb-20 22:08:36 - 2 of 18 assignments: Endo\n",
      "19-Feb-20 22:08:36 - 3 of 18 assignments: L2/3 IT\n",
      "19-Feb-20 22:08:36 - 4 of 18 assignments: L5 ET\n",
      "19-Feb-20 22:08:36 - 5 of 18 assignments: L5 IT\n",
      "19-Feb-20 22:08:36 - 6 of 18 assignments: L5/6 NP\n",
      "19-Feb-20 22:08:37 - 7 of 18 assignments: L6 CT\n",
      "19-Feb-20 22:08:37 - 8 of 18 assignments: L6 IT\n",
      "19-Feb-20 22:08:37 - 9 of 18 assignments: L6 IT Car3\n",
      "19-Feb-20 22:08:37 - 10 of 18 assignments: L6b\n",
      "19-Feb-20 22:08:37 - 11 of 18 assignments: Lamp5\n",
      "19-Feb-20 22:08:37 - 12 of 18 assignments: Low Quality\n",
      "19-Feb-20 22:08:37 - 13 of 18 assignments: Pvalb\n",
      "19-Feb-20 22:08:37 - 14 of 18 assignments: SMC\n",
      "19-Feb-20 22:08:37 - 15 of 18 assignments: Sncg\n",
      "19-Feb-20 22:08:37 - 16 of 18 assignments: Sst\n",
      "19-Feb-20 22:08:38 - 17 of 18 assignments: VLMC\n",
      "19-Feb-20 22:08:38 - 18 of 18 assignments: Vip\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.8 s, sys: 11.1 ms, total: 1.82 s\n",
      "Wall time: 1.81 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "mat         = same.layers[\"log1p\"].todense()\n",
    "components  = same.obs.cell_id.values\n",
    "features    = same.var.TSS_class.values\n",
    "assignments = same.obs.subclass_label.values\n",
    "\n",
    "unique_features = np.unique(features)\n",
    "\n",
    "z = deSS(mat, components, features, assignments, nan_cutoff=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 581,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([41428])"
      ]
     },
     "execution_count": 581,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_features[z[-6]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 582,
   "metadata": {},
   "outputs": [],
   "source": [
    "interesting = 41428"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 589,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>new_start</th>\n",
       "      <th>TSS_same</th>\n",
       "      <th>n_TSS_same</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tss_class</th>\n",
       "      <th>n_iso</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>44668</th>\n",
       "      <td>17</td>\n",
       "      <td>ENSMUSG00000090417.3</td>\n",
       "      <td>18277756</td>\n",
       "      <td>[ENSMUST00000231815.1, ENSMUST00000172190.3, E...</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUST00000231815.1</td>\n",
       "      <td>41428</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome               gene_id  new_start  \\\n",
       "44668         17  ENSMUSG00000090417.3   18277756   \n",
       "\n",
       "                                                TSS_same  n_TSS_same  \\\n",
       "44668  [ENSMUST00000231815.1, ENSMUST00000172190.3, E...           4   \n",
       "\n",
       "              transcript_id  tss_class  n_iso  \n",
       "44668  ENSMUST00000231815.1      41428      4  "
      ]
     },
     "execution_count": 589,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "same_t2g[same_t2g[\"tss_class\"] == 41428]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 583,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>gene_TPM</th>\n",
       "      <th>TSS_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>81241</th>\n",
       "      <td>ENSMUST00000231815</td>\n",
       "      <td>ENSMUSG00000090417</td>\n",
       "      <td>Vmn2r94</td>\n",
       "      <td>Vmn2r94-204</td>\n",
       "      <td>8634.106445</td>\n",
       "      <td>41428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81242</th>\n",
       "      <td>ENSMUST00000231390</td>\n",
       "      <td>ENSMUSG00000090417</td>\n",
       "      <td>Vmn2r94</td>\n",
       "      <td>Vmn2r94-202</td>\n",
       "      <td>6814.746094</td>\n",
       "      <td>41428</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id             gene_id gene_name transcript_name  \\\n",
       "81241  ENSMUST00000231815  ENSMUSG00000090417   Vmn2r94     Vmn2r94-204   \n",
       "81242  ENSMUST00000231390  ENSMUSG00000090417   Vmn2r94     Vmn2r94-202   \n",
       "\n",
       "          gene_TPM  TSS_class  \n",
       "81241  8634.106445      41428  \n",
       "81242  6814.746094      41428  "
      ]
     },
     "execution_count": 583,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "same.var[same.var.TSS_class==interesting]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 584,
   "metadata": {},
   "outputs": [],
   "source": [
    "u = interesting\n",
    "\n",
    "a = np.asarray(tmp1[:,tmp1.var.TSS_class == u].layers[\"log1p\"].todense())\n",
    "b = np.asarray(tmp2[:,tmp2.var.TSS_class == u].layers[\"log1p\"].todense())\n",
    "\n",
    "y = normalize_rows_L1(a)\n",
    "yy = normalize_rows_L1(b)\n",
    "\n",
    "y = y[~np.isnan(y[:,0])]\n",
    "yy = yy[~np.isnan(yy[:,0])]\n",
    "\n",
    "yerr = sem(y, axis=0)\n",
    "yyerr = sem(yy, axis=0)\n",
    "\n",
    "y = y.mean(axis=0)\n",
    "yy = yy.mean(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 586,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "19-Feb-20 22:11:53 - update_title_pos\n",
      "19-Feb-20 22:11:53 - update_title_pos\n",
      "19-Feb-20 22:11:53 - update_title_pos\n",
      "19-Feb-20 22:11:53 - update_title_pos\n",
      "19-Feb-20 22:11:53 - update_title_pos\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyYAAAIGCAYAAACh7JuwAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAWJQAAFiUBSVIk8AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3X24plddH/rvL0xIhglNIoS8EUnCS5IDESQJLyKCRxBRBg1QLOC5BA8UFClWaw02aChWpFipkB5Q4CrnWDhWhUGophRaCdgGmQwHSCAgkgAJL2MQQoRMMCG/88ezhzz7nj0z+5l977nn5fO5rufKXutZa92/RK7tfGfd972quwMAADClI6YuAAAAQDABAAAmJ5gAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJMTTAAAgMkJJgAAwOQ2TF3AVKrqe5Kck+TeSb6d5IYkH+3uT09aGAAAHIaqu8dftOqIzP7Q/7AkFyx9vifJXeeGPae73zz6xfde2/OSvDjJA3cz5K+SvLK7t+y/qgAA4PA26o5JVT0tyc8nOS/JMWOuvVZVdVySP0jypL0MfXiSt1fVG5K8sLtvW/fiAADgMDf2rVzfn+QxI6+5ZlV1lyR/lOTxg68+k+SqJEcm+d4kp8x997ylf/7TdS8QAAAOc/vr4fevJ/nCfrrWSn4jy0PJ15NcmOT+3X1hdz8pyX2SPD/JP8yNe15V/cz+KxMAAA5P6xFMdiS5IslrkvwfSc5OcnySN67Dtfaqqk5N8gtzXbcleVx3v6PnHrDp7tu7+/eT/JPBEi+vqqP3Q6kAAHDYGvtWrn+T5F909+3DL6pq5Eut2kVJ5oPFq7v7yt0N7u4tVfXHSf7xUtcpme2k/O76lQgAAIe3UXdMuvvGlULJVJbeDvaTc13fzuoCxu8M2s8arSgAAGAXh/oBi49IcsJc+y+6+4t7m9TdH8zswfidzq+qk8cuDgAAmDnUg8nwLVwfWGDuX879XEket/ZyAACAlRzqwWR4iOKHFpj7wb2sBQAAjORQDyZnDdrXLjD3ur2sBQAAjORQDyb3HbSvX2DuDYP2/dZYCwAAsBuHbDBZOu1901zXN7t7xwJL3DhoH7v2qgAAgJUcssEkyTGD9q0Lzh+GmOF6AADASMY+YPFAsmnQXjSYDMcP19utqtq2m68elOQbST67YC0AAHAgOj3Jzd19xloXOpSDyVCvcfwYR9ffZePGjd91zjnnfNcIawEAwKSuueaa7NixyNMSu3coB5NvDtobF5w/HP+N1U7s7vNW6q+qbeecc85Dt23b3YYKAAAcPM4777x8+MMf/uwYax3Kz5gMg8nRC84fjl91MAEAABZzyAaT7r49yS1zXZuqapFdkxMG7a+vvSoAAGAlh2wwWfKZQfveC8w9bS9rAQAAIznUg8knB+0zF5g7fLPAcC0AAGAkh3ow+fig/fAF5g7HfmKNtQAAALtxqAeT9w7aj15g7vzYXmEtAABgJId6MLkiyY1z7R+sqlP2NqmqHpHkvnNdV3b3F8cuDgAAmDmkg0l335HkP8913SXJi1cx9RcH7beMVhQAALCLgyqYVNVjq6rnPp9dxbTfSvKtufY/r6rz93CNn0jyj+e6vpTk9/epYAAAYFVGP/m9qk7fzVfHDdr33M3YW7v7y2PV091fqKpXJ7loqevIJO+pqmd395/uHFdVG5I8J8mlgyVe2t07xqoHAADY1ejBJMl1qxz3qqXP0OVJHjtaNTMvTXJ+kscttY9L8o6q+pskV2UWVr43yamDeW/s7jeNXAsAADCwHsHkgNPdt1fVP07yn5L82NxX91v6rOSNSX5uvWsDAAAOsmdM1qK7b+ruJyX5p9nzmSQfSvKU7n5ed9+2f6oDAIDD2+g7Jt1dY685t/b7kqxp/e5+Q5I3VNVDkpyT2e1b305yQ5KPdvdfr7VOAABgMYfFrVwr6e6PJPnI1HUAAACH0a1cAADAgUswAQAAJieYAAAAkxNMAACAyQkmAADA5AQTAABgcoIJAAAwOcEEAACYnGACAABMTjABAAAmJ5gAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJMTTAAAgMkJJgAAwOQEEwAAYHKCCQAAMDnBBAAAmJxgAgAATE4wAQAAJieYAAAAkxNMAACAyQkmAADA5AQTAABgcoIJAAAwOcEEAACYnGACAABMTjABAAAmJ5gAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJMTTAAAgMkJJgAAwOQEEwAAYHKCCQAAMDnBBAAAmJxgAgAATE4wAQAAJieYAAAAk9uw3heoqg1Jvi/J6UlOTnJzkhuSXNHdX1nv6++mpnsmeViS705yfJLbk3wtyaeSXNndO6aoCwAADlfrFkyq6m5JXprkOUlOXGHIbVV1WZKLu/uq9apjUNNPJPnnSX5gD8P+oar+JMkru/tj+6MuAAA43K3LrVxV9cAk25JclJVDSZIcmeTJST5UVS9Yjzrm6tlUVW9LsiV7DiVJctckz0xyZVX98nrWBQAAzIy+Y1JVJyd5d5JTB19tS3JtknskuSDJ3Zf6j07yuqq6ubvfug71bEjyziT/++CrW5NsTfKFzELS/ZOcm6SWvj8yyb+tqnT3q8auCwAAuNOoOyZVVUneluWh5KokD+7u87v76d39Q5k923HpYPqblnZaxvaiLA8lneS3k5zc3T/Q3c/o7qd194OTnJPkvw/m/2ZVnbMOdQEAAEvGvpXrKUkeOde+Lsljhs9qdPdN3f2iJK+Z6z46ycvHLGYpKP3SoPvXu/uXu/um4fju/lSSH0nyvrnuDUn+xZh1AQAAy40dTH590H5hd39tD+NfkuRzc+0Lq+ohI9bzPVm+e/OlJK/c04Tuvj3JLwy6f3TEmgAAgIHRgklVnZvZMxo7XdPdl+1pTnffkuT1g+5njlVTZq8onvee7v6HvU3q7o9m9krjnU5aessYAACwDsbcMdk8aL9llfOG4548Qi07bRq0b1hx1MquH7SPW2MtAADAbowZTB4/aH9gNZO6+/osv53rrKo6baSatg/aGxeYOz+2k+zyTAoAADCOMYPJ/Bu17khy5QJzP7iHtdbiyiTzt26t6vmVqtqY5Oy5rk8t3XYGAACsg1GCSVUdn+SEua7tC/5B/rpB+6y1V5V099eT/Oe5rses8pXEz83sLWE7rfa2NAAAYB+MtWNy30F7+HzG3gyf/bjfGmoZ+pXceUvXEUn+uKqGhz9+R1U9Lsvf3PX5JL87Yj0AAMDAWCe/Hzto37jg/OH44Xr7rLu/tBQ2/jTJmZkdonhVVb0+yXuy/OT3pyX5J0nusjT9b5M8qbv/fqx6AACAXY0VTI4ZtG9dcP6Ovay3Jt19dVU9OLNT4J+TWQh5ydJnJd9O8kdJ/kV3f3HMWgAAgF2NFUyGr+VdNJgMxw/XG8POXZBv7WXcHUkuTfLqfQ0lVbVtN1+dvZt+AAA4rI198vtOvcbxNVYhSVJVP5rk00l+M8mD9jL8iCQvTvI3VfXapTd0AQAA62isHZNvDtqL/mF+OP4ba6hlmaq6MMkf584dkyTZmuS1mZ218qXMnjE5I8mPJPmFJKdk9t/m55OcX1U/vMhzJt193m5q2ZbkofvwrwEAAIe0sXZMhsHk6BVH7d5w/CjBZOntW/8xy0PJv03yiO7+g+7+bHd/q7u/0d1XdferMjtD5X1z4x+R5PfGqAcAAFjZWMHk64P2PRecf8KgPVxvX/3zLH/D159396909x27m9DdNyW5MMtPjX9GVV0wUk0AAMDAWMHkM4P2aQvOH44frrevnjpov2I1k5bCyX8YdP/UKBUBAAC7GCWYdPdXs/wskpOq6m4LLHHGoP3JtdZUVcckOX2u61tJrlhgifcN2nZMAABgnYz5Vq6PD9Y9f4G5Dx+0P7H2cnY5pPHvuvvbC8zfPmgvensaAACwSmMGk/cO2o9ezaSqOi3LdzY+1d2fH6GemwftRc9GGY4f7U1hAADAcmMGk3cO2s9a5bzhuOE6+2Tp9b7zYeLYqvruBZZ4yKA93EEBAABGMlow6e6rklw913VOVT1xT3OWDi98waD7rWPVlNk5JfOes8Dc4dj/ucZaAACA3Rj75PdLBu1Lq+r4PYx/RZL7zLXf0d0f2d3gqnp2VfXc5317qedtg/avrOa1v1X14iy/Fa2T/One5gEAAPtm7GDy9ix/89WZSS6vqnPnB1XVsVX12iQvnuu+NcnFI9fz/2T5q4c3JvmLqnphVe1yCGRVnVRV/yHJvx989cdLO0IAAMA62DDmYt3dVfW0JFuTnLLUfW6Sj1bVtiTXJrlHkocluftg+nO7++MZUXffVlU/meTy3Pkw+6YklyZ5ZVV9KMmXkxyZ2SuLH5Llp8QnyaeT/PyYdQEAAMuNGkySpLu/WFVPSPInSc5a6q7MXh+80iuEb03yi939lrFrWapnW1U9PrNnV06f+2pTkh/cy/S/TPKs7r5xL+MAAIA1GPtWriRJd1+d5KFJXpnkb3cz7LbM3sD1sO5+3XrUMVfPFUm+J8m/zOpOlf+rJD+d5DEjvboYAADYg9F3THbq7luSXFRVFyd5VGa3Sp2U2fkiNyS5YtGdiO5+c5I372M9f5/kVUletfTa4POX6jk2yR1JbkryuSRbu/vv9uUaAADAvlm3YLJTd9+e2TMel6/3tVZraRfETggAABwg1uVWLgAAgEUIJgAAwOQEEwAAYHKCCQAAMDnBBAAAmJxgAgAATE4wAQAAJieYAAAAkxNMAACAyQkmAADA5AQTAABgcoIJAAAwOcEEAACYnGACAABMTjABAAAmJ5gAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJMTTAAAgMkJJgAAwOQEEwAAYHKCCQAAMDnBBAAAmJxgAgAATE4wAQAAJieYAAAAkxNMAACAyQkmAADA5AQTAABgcoIJAAAwOcEEAACYnGACAABMTjABAAAmJ5gAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADC5DVMXAPvTJZdcsuLPAABMSzDhsPKyl73sOz8LJgAABw63cgEAAJMTTAAAgMkJJhw2brnllmXtHTt2TFQJAABDggmHha1bt+bMM89c1nfGGWdk69atE1UEAMA8wYRD3o4dO7J58+Zs3759Wf/27duzefNmOycAAAeAdX8rV1VtSPJ9SU5PcnKSm5PckOSK7v7Kel9/T6pqY5KHJzk7yfFL3TcluTbJR7v7y1PVxni2bNmySyjZafv27dmyZUue+cxn7ueqAACYt27BpKruluSlSZ6T5MQVhtxWVZclubi7r1qvOnZT24OT/EqSC5McvYdxn0nyZ0ku6u5D4q/VT7/oz6YuYb+76X9dtsfvf+71l+VXP3bsfqrmwPDZ3/qxqUsAAFhmXW7lqqoHJtmW5KKsHEqS5MgkT07yoap6wXrUsUJdR1bVv03y4STPyB5CyZL7JvlnSQ6vP7UeYo487qQ1fQ8AwPobfcekqk5O8u4kpw6+2pbZLVL3SHJBkrsv9R+d5HVVdXN3v3XseubqOjrJ25M8cfDVtzILKl9e+vmeSR6UxJ9WDxEb7//IHLHpuNzxzZt2+e6ITcdl4/0fOUFVAADMG3XHpKoqyduyPJRcleTB3X1+dz+9u38oyXcnuXQw/U1LOy3r5c1ZHkr+Nslzk9yzu7+vu5/S3c/o7sd398lJHpDkJZmFKQ5iRxx5VO711F/LEZuOW96/6bhZ/5FHTVQZAAA7jX0r11OSzP/183VJHtPdH5sf1N03dfeLkrxmrvvoJC8fuZ4kSVU9M8lPznX9f0nO6e43dfc3VprT3Z/u7t9Kcv/MQgwHsaNOfkBOff6blvWd+vw35aiTHzBRRQAAzBs7mPz6oP3C7v7aHsa/JMnn5toXVtVDxiyoqo5P8jtzXV9O8vju/upq5nf3Hd19x5g1MY3hzoidEgCAA8dowaSqzk1y7lzXNd29x9chdfctSV4/6B77va0/l+UP4F/U3X838jUAAIA1GHPHZPOg/ZZVzhuOe/IItST5zjMvPzPX9ZUVrgcAAExszGDy+EH7A6uZ1N3XZ/ntXGdV1Wkj1fToJGfOtd/e3bePtDYAADCSMV8XPP9GrTuSXLnA3A8muc9gretHqOkHVrgOh7FjH/WMqUsAAGAFowSTpQfMT5jr2r70/MhqXTdon5Xkv665sOT8QfvqJKmqu2Z26vtPZRaCTknyjcwejP/LJO/o7jGuzwHmuO9/1tQlAACwgrF2TO47aC+623HDoH2/NdQy73uH16mqB2X2nMn3DL47KrPDHx+Y5PlV9cEkz+3uj49UCwAAsBtjPWNy7KB944Lzh+OH6+2r4entZyT5X9k1lKzkEUn+Z1UNbwcDAABGNlYwOWbQvnXB+Tv2st7CqmpjkrsOuv9zkrsv/fzJJC9KckGSszN7eP+1Sf5hbvyxSd5eVfMn2QMAACMb61auTYP2osFkOH643r5Yadfl3kv//IPMbtOaDyGfSvLeqnpjkvckuddS/z2SXJrZMymrUlXbdvPV2atdAwAADidjn/y+U69xfI1Qw+7+3a5M8pxBKLmzkO6PJXn6oKYfr6oHjFATAACwgrF2TL45aG9ccP5w/DfWUMtOw5p2ekl3f3tPE7v78qr6syRPWuqqJD+Z5OWruXB3n7dS/9JOykNXswYAABxOxtoxGYaAoxecPxw/RjBZaY0bk/z3Vc5/66D9/WsrBwAA2J2xgsnXB+17Ljj/hEF7uN7ClnZFhoFpW3ev9jazrYP2WWutCQAAWNlYweQzg/ZpC84fjh+ut68+PWh/eYG5w7HftcZaAACA3RglmHT3V7P8LJKTqupuCyxxxqD9ybVXlSS5ZtD+1gJzh2OPWmMtAADAboz5Vq75E9KPSHL+AnMfPmh/Yu3lJEmuHrQXObjxuEH7q2usBQAA2I0xg8l7B+1Hr2ZSVZ2W5PS5rk919+dHqundg/Y5C8wdjv3iGmsBAAB2Y8xg8s5B+1mrnDccN1xnn3X3tiTzIedBVXXSKqc/btD+n+NUBQAADI0WTLr7qiy/deqcqnrinuZU1cYkLxh0D1/Tu1b/ae7nuyT5ub1NqKpNSf7PQfefj1kUAABwp7FPfr9k0L60qo7fw/hXJLnPXPsd3f2R3Q2uqmdXVc993reKmn47yU1z7X9ZVRfsZc5rkpwy174qu94WBgAAjGTsYPL2JFfMtc9McnlVnTs/qKqOrarXJnnxXPetSS4euZ5099eyPDAdleS/VdVPVdWyf/+quldVvTXJz8x135HkFxc4/wQAAFjQhjEX6+6uqqdldjjhzh2Hc5N8tKq2Jbk2yT2SPCzJ3QfTn9vdH8866O7frarvTfLTS13HJfmDJL9dVVuT/H2Seyd5ZHb9b/LS7h4+2A8AAIxo1GCSJN39xap6QpI/yZ2npVdmrw9e6RXCt2a2I/GWsWsZeN7StZ4/13dikiftZvy3k/xCd1+6znUBAMBhb+xbuZIk3X11kocmeWWSv93NsNsyewPXw7r7detRx6Cm27r7BUl+NMn7k+zu1qxvJfmjJA8SSgAAYP8Yfcdkp+6+JclFVXVxkkdldrr7SUluTnJDkiu6+8Y9LLHSmm9O8uY11nVZksuWzk85P8mpmd1W9ndJPpfkA0u1AwAA+8m6BZOduvv2JJcvfQ4Y3X19kuunrgMAAFinW7kAAAAWIZgAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJPbMHUBAPvTLbfcki1btuS6667LmWeemQsvvDAbN26cuiwAOOwJJsBhY+vWrdm8eXO2b9/+nb4TTzwx73rXu3LBBRdMWBkA4FYu4LCwY8eOXUJJkmzfvj2bN2/Ojh07JqoMAEgEE+AwsWXLll1CyU7bt2/Pli1b9nNFAMA8t3LB4eiSY6euYL+79v3f2vP3b/yZ5K9/dj9Vc4C45OtTVwAA32HHBDgsnHn8nn/d7e17AGB9+f/EwGHhwrM35MRNteJ3J26qXHi2DWQAmJJgAhwWNh5Zedcz7rZLODlx06x/45ErhxYAYP/wV4TAYeOCU++S6158TLZ88vZc+7U7cubxR+TCszcIJQBwABBMgMPKxiMrzzz3yKnLAAAG3MoFAABMTjABAAAmJ5gAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJPbMHUBAACwL2655ZZs2bIl1113Xc4888xceOGF2bhx49RlsY8EEwAADjpbt27N5s2bs3379u/0nXjiiXnXu96VCy64YMLK2Fdu5QIA4KCyY8eOXUJJkmzfvj2bN2/Ojh07JqqMtRBMAAA4qGzZsmWXULLT9u3bs2XLlv1cEWNwKxcAwMHqkmOnrmAS177/W3v+/o0/k/z1z+6nag4Ql3x96grWzI4JAAAHlTOP3/MfYff2PQcm/1cDAOCgcuHZG3LiplrxuxM3VS48201BByPBBACAg8rGIyvvesbddgknJ26a9W88cuXQwoFNnAQA4KBzwal3yXUvPiZbPnl7rv3aHTnz+CNy4dkbhJKDmGACAMBBaeORlWeee+TUZTASt3IBAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJMTTAAAgMltWO8LVNWGJN+X5PQkJye5OckNSa7o7q+s9/UBAIAD37oFk6q6W5KXJnlOkhNXGHJbVV2W5OLuvmq96liNqjozydVJNg6+OqO7P7v/KwIAgMPLutzKVVUPTLItyUVZOZQkyZFJnpzkQ1X1gvWoYwGvy66hBAAA2E9G3zGpqpOTvDvJqYOvtiW5Nsk9klyQ5O5L/UcneV1V3dzdbx27nr2pqmcl+eH9fV0AAOBOo+6YVFUleVuWh5Krkjy4u8/v7qd39w8l+e4klw6mv2lpp2W/qarvSvLqua5v7M/rAwAAM2PfyvWUJI+ca1+X5DHd/bH5Qd19U3e/KMlr5rqPTvLykevZm1clOWHp53dmtqsDAADsZ2MHk18ftF/Y3V/bw/iXJPncXPvCqnrIyDWtqKoek9mD+UlyS5J/tj+uCwAA7Gq0YFJV5yY5d67rmu6+bE9zuvuWJK8fdD9zrJp2p6qOSvJ7SWqp61939+f2MAUAAFhHY+6YbB6037LKecNxTx6hlr15SZKzln7+RJLf2Q/XBAAAdmPMYPL4QfsDq5nU3ddn+e1cZ1XVaaNVNVBVZ2cWTHb62e6+bb2uBwAA7N2YwWT+jVp3JLlygbkf3MNao1l6a9jvJbnrUtf/3d3vX49rAQAAqzdKMKmq43Pn262SZPvS8yOrdd2gfdaKo9buZ5L8wNLPX0vyy+t0HQAAYAFj7Zjcd9C+fsH5Nwza91tDLSuqqntl9nrgnS7q7hvHvg4AALC4sYLJsYP2on/gH44frjeGVyc5funnDyZ5wzpcAwAA2AdjBZNjBu1bF5y/Yy/rrUlV/XDufA3xtzN74L3HvAYAALDvNoy0zqZBe9FgMhw/XG+fVdXGJK+b63ptd39krPV3c83dnSB/9npeFwAADlZjn/y+06K7EcPxteKoffPrSc5c+vmLSX5txLUBAIARjLVj8s1Be+OC84fjv7GGWr5j6TT6X5rr+oXu/vsx1t6T7j5vN/VsS/LQ9b4+AAAcbMbaMRkGk6MXnD8cv+ZgUlVHJPn93Bm+3t3df7zWdQEAgPGNFUy+Pmjfc8H5Jwzaw/X2xc8mecTSz7cmeeEIawIAAOtgrFu5PjNon7bg/OH44Xr74uK5n9+c5NtVdfpe5gx3bu49Oyz+O/52wYMjAQCAVRglmHT3V6vqxty583FSVd1tgT/EnzFof3KEsuafW3nB0mdRHxi0L0zyjn2uCAAYxSWXXLLiz8DBa6wdkyT5eJLHLv18RJLzk7x/lXMfPmh/YqSaAIBD0Mte9rLv/CyYwKFhzNcFv3fQfvRqJlXVaUlOn+v6VHd/fqyiAACAA9+YweSdg/azVjlvOG64zj7p7uO6uxb5JLl8sMwZgzFu4wIAgHUwWjDp7quSXD3XdU5VPXFPc5ZOZR8++/HWsWoCAAAODmOf/H7JoH1pVR2/h/GvSHKfufY7uvsjuxtcVc+uqp77vG/fSwUADka33LL83To7duyYqBJgTGMHk7cnuWKufWaSy5dOYP+Oqjq2ql6b5MVz3bdm+St+AQCW2bp1a84888xlfWeccUa2bt06UUXAWMZ8K1e6u6vqaUm2JjllqfvcJB+tqm1Jrk1yjyQPS3L3wfTndvfHx6wHADh07NixI5s3b8727duX9W/fvj2bN2/Oddddl40bN+5mNnCgG3vHJN39xSRPSPKpue7K7PXBT0/yQ1keSm5N8nPd/ZaxawEADh1btmzZJZTstH379mzZsmU/VwSMadQdk526++qqemiSX0vynCT3WmHYbUkuS3Lx0oPzAMAqnX7Rn01dwn530/+6bI/f/9zrL8uvfuzY/VTNgeGzR09dAYxnXYJJkiyd+n5RVV2c5FGZne5+UpKbk9yQ5IruvnHBNd+c5M3jVrps/ceu19oAwNocedxJa/oeOLCtWzDZqbtvz+x8kOEZIQAAq7bx/o/MEZuOyx3fvGmX747YdFw23v+RE1QFjGX0Z0wAANbDEUcelXs99ddyxKbjlvdvOm7Wf+RRE1UGjEEwAQAOGked/ICc+vw3Les79flvylEnP2CiioCxCCYAwEFluDNipwQODYIJAAAwOcEEAACY3Lq/lQsAYGzHPuoZU5cAjEwwAQAOOsd9/7OmLgEYmVu5AACAyQkmAADA5AQTAABgcoIJAAAwOcEEAACYnGACAABMTjABAAAmJ5gAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJMTTAAAgMkJJgAAwOQEEwAAYHKL5YbSAAASI0lEQVSCCQAAMDnBBAAAmJxgAgAATE4wAQAAJieYAAAAkxNMAACAyQkmAADA5AQTAABgcoIJAAAwOcEEAACYnGACAABMTjABAAAmJ5gAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJMTTAAAgMkJJgAAwOQEEwAAYHKCCQAAMLkN632BqtqQ5PuSnJ7k5CQ3J7khyRXd/ZX1vv6glqOTnJPkf0tyQpJNS/XcmGRbd396f9YDAADMrFswqaq7JXlpkuckOXGFIbdV1WVJLu7uq9axjjOTPD3JD2cWkI7aw9gvJHl9kv/Q3V9br5oAAIDl1uVWrqp6YJJtSS7KyqEkSY5M8uQkH6qqF6xTHX+Y5DNJXpHkB7OHULLk1CQvT/Lxqvrh9agJAADY1eg7JlV1cpJ3Z/aH/Hnbklyb5B5JLkhy96X+o5O8rqpu7u63jlzO/Vbo6ySfSHJ9kq8mOS7J+UnuNTfm5CR/VlUXdvd/GbkmAABgYNRgUlWV5G1ZHkquSvJT3f2xuXHHZbYz8fNz495UVR/t7o+PWdOSTvI/krwxyXu6++9WqPsnkrx2rvYNSf6wqs7u7hvWoSYAAGDJ2LdyPSXJI+fa1yV5zHwoSZLuvqm7X5TkNXPdR2cWVsZ0R5I/TnJOdz+uu/9wGEqW6unu3pLZTs7n5r7atA41AQAAA2MHk18ftF+4l4fIX5LlQeDCqnrIiPU8rbuf3t2fWs3g7v5SkucOup9aVXcdsSYAAGBgtGBSVecmOXeu65ruvmxPc7r7lszegjXvmWPV1N2f34c5781sp2enuyf53rFqAgAAdjXmjsnmQfstq5w3HPfkEWpZq48O2qdMUgUAABwmxgwmjx+0P7CaSd19fZbfznVWVZ02WlX75vZB+8hJqgAAgMPEmMHkgXM/35HkygXmfnAPa03hvoP29kmqAACAw8QowaSqjk9ywlzX9qXnR1brukH7rLVXtW+q6j5J5h/Avz273toFAACMaKwdk+EOw/ULzh+eE7LSwYj7y88nqbn2B7r7pqmKAQCAw8FYweTYQfvGBecPxw/X2y+q6pwkLxp0//spagEAgMPJWMHkmEH71gXn79jLeuuuqo5K8tYkR811//fufuf+rgUAAA43G0ZaZ9OgvWgwGY4frrc/vD7Lny35+yTP25eFqmrbbr46e1/WAwCAQ93YJ7/v1GscXyuOWidV9S+TPHuuq5M8t7uHD+UDAADrYKwdk28O2hsXnD8c/4011LKQqvqpJL816P6V7v6jfV2zu8/bzbW2JXnovq4LAACHqrF2TIbB5OgF5w/H75dgUlU/muQ/ZvkOzSu7+1X74/oAAMDMWMHk64P2PRecf8KgPVxvdFX1/Un+JMt3jd7Q3Ret97UBAIDlxgomnxm0T1tw/nD8cL1RVdX3JvkvWX4L2R8lecF6XhcAAFjZKMGku7+a5WeRnFRVd1tgiTMG7U+uvaqVVdVZSd6d5WelXJbkp7r7jvW6LgAAsHtjvpXr44N1z19g7sMH7U+svZxdVdV3J3lPlt869oEkT+3u29bjmgAAwN6NGUzeO2g/ejWTquq0JKfPdX2quz8/VlFz17lXZjXO3za2LcmTunt4wCMAALAfjRlMhiekP2uV84bjRj9pvaqOS/Lfktx/rvuaJD/S3TePfT0AAGAxowWT7r4qydVzXedU1RP3NKeqNmbXB87fOlZNc9d4V5IHz3Vfl+Tx3f2VMa8FAADsm7FPfr9k0L60qo7fw/hXJLnPXPsd3f2R3Q2uqmdXVc993renYqrqyCRvS/L9c91fTPK47v7CnuYCAAD7z1gnv+/09iRXJHnkUvvMJJdX1bOWdlSSJFV1bJLfSPLzc3NvTXLxyPW8Ocn8rs2OJM9LckdVnb7AOjd1903jlQUAAMwbNZh0d1fV05JsTXLKUve5ST5aVduSXJvkHkkeluTug+nP7e6PZ1zPHLQ3JvmzfVjnZdl1NwgAABjJ2Dsm6e4vVtUTMjtV/ayl7srs9cErvUL41iS/2N1vGbsWAADg4DD2MyZJku6+OslDk7wyyd/uZthtmb2B62Hd/br1qAMAADg4jL5jslN335Lkoqq6OMmjMjvd/aQkNye5IckV3X3jHpZYac03Z/bcyGrH1yLrAwAA01i3YLJTd9+e5PKlDwAAwC7W5VYuAACARQgmAADA5AQTAABgcoIJAAAwOcEEAACYnGACAABMTjABAAAmJ5gAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJMTTAAAgMkJJgAAwOQEEwAAYHKCCQAAMDnBBAAAmJxgAgAATE4wAQAAJieYAAAAkxNMAACAyQkmAADA5AQTAABgcoIJAAAwOcEEAACYnGACAABMTjABAAAmJ5gAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJMTTAAAgMkJJgAAwOQEEwAAYHKCCQAAMDnBBAAAmJxgAgAATE4wAQAAJieYAAAAkxNMAACAyQkmAADA5AQTAABgchvW+wJVtSHJ9yU5PcnJSW5OckOSK7r7K+t9/d3UVEkeluR+SU5NcstSTdu6+/opagIAgMPZugWTqrpbkpcmeU6SE1cYcltVXZbk4u6+ar3qGNS0IckvJXlBZkFp6I6q+osk/7q7378/agIAANbpVq6qemCSbUkuysqhJEmOTPLkJB+qqhesRx2Dmu6d5ANJfisrh5Jk9t/jh5L8RVX96/WuCQAAmBl9x6SqTk7y7sxukZq3Lcm1Se6R5IIkd1/qPzrJ66rq5u5+69j1LNV0TJI/T3Lu4KuPJ/nkUi3nLdWWzALKS6vq1u7+zfWoCQAAuNOoOyZLz268LctDyVVJHtzd53f307v7h5J8d5JLB9PftLTTsh7ekOWh5IYkP9DdD+rup3X3E5KcltmtZz037jeq6nHrVBMAALBk7Fu5npLkkXPt65I8prs/Nj+ou2/q7hclec1c99FJXj5yPamq85P8k7mum5I8urs/MKhpR3f/RmbPoHxnepJXjl0TAACw3NjB5NcH7Rd299f2MP4lST43176wqh6yzjX9and/dg/j/32Sv5prP7SqfnzkmgAAgDmjBZOqOjfLb5e6prsv29Oc7r4lyesH3c8csabjk/zIXNfXkvzHvdTUmYWTec8aqyYAAGBXY+6YbB6037LKecNxTx6hlp2emOUP+L+tu29dxbx3ZHa2yU5PqKq7jlgXAAAwZ8xg8vhB+wMrjhpYOtBw/naus6rqtIlrujXJ1rmuf5Tk4SPVBAAADIwZTObfqHVHkisXmPvBPay1FsN1PrTA3PWqCQAAGBglmCw9y3HCXNf2pedHVuu6QfustVe1yzq9wnX2ZL1qAgAABsbaMbnvoH39gvNvGLTvt4ZakiRVdc/MbsHa6cbu/taUNQEAACsbK5gcO2jfuOD84fjhevviQKwJAABYwVjB5JhBezVvvpq3Yy/r7YsDsSYAAGAFG/Y+ZFU2DdqLhoDh+OF6+2Kymqpq226+evA111yT8847b8FSxvOlL3x9smtz4DjviG9MXQIHgndN97uItfP7nMTvc+ZM9Dv9mmuuSZLTx1hrrGAy1GscX2MVsodrLDp+jJq+vWPHjq9/+MMf/uwIa8G+OvvDs39+ctoymNyXPjx1BcDa+H3Onab7nX56kpvHWGisYPLNQXvjgvOH48eI/5PV1N3+GpID1s4dPf87BTi4+X3OoWasZ0yGIeDoBecPx69HMDkQagIAAFYwVjAZ3uh6zwXnnzBoj3Hj7IFYEwAAsIKxgslnBu3TFpw/HD9cb2HdfWOW3+92r6q665Q1AQAAKxslmHT3V7P83I+TqupuCyxxxqA91kNcn5r7uVa4zp6sV00AAMDAWDsmSfLxwbrnLzD34YP2J9ZeTpLlNa10nT1Zr5oAAICBMYPJewftR69mUlWdluXvPv5Ud39+4pqOTnLBXNffJ/ngSDXBpLr7PG9wATj4+X3OoWbMYPLOQftZq5w3HDdcZy3+PMntc+2nLoWOvfmJLD9Q8b929z+MWBcAADBntGDS3VcluXqu65yqeuKe5lTVxiQvGHS/dcSavpbkv851HZ/kOXupqZL8wqD7LWPVBAAA7GrMHZMkuWTQvrSqjt/D+Fckuc9c+x3d/ZHdDa6qZ1dVz33et4qaXjZo/2ZV3WfFkTMvzvLnSz6ccXdxAACAgbGDyduTXDHXPjPJ5VV17vygqjq2ql6bWQjY6dYkF49cT7r7yiR/ONd1XJK/rKplz5tU1dFV9a+S/M789CS/0t09dl0AAMCdNoy5WHd3VT0tydYkpyx1n5vko1W1Lcm1Se6R5GFJ7j6Y/tzuHr5FayzPS/LApVqS5N5J3l9VV2f2GuBjMnuL2PAQxpd29/ABegAAYGSjBpMk6e4vVtUTkvxJkrOWuiuzP/iv9ArhW5P8Ynev23Mc3f2NqvqxJH+U5BFzXz1o6TN0R5Lf7O5/s141AQAAdxo9mCRJd19dVQ9N8muZPWx+rxWG3ZbksiQXLz04v666+/ql27d+KcnPZvmzLd8ZluQvkrysu9+/3jUBAKyHqjoiyf0yu2Pk1CT/KMmOJF/N7Jy3j3T37btfAfa/Wu/HJ6pqQ5JHZXaS+klJbk5yQ5IruvvGPc1dx5oqswfc75fZLWc7knwhydbuvn6KmmCnqjo9yXXrfJmXdfclK1z7xMzO8HnY0j8vyOz2y50u7+7HrnNtAIeE/f37vKq+K8mFSX40yQ9m9jbS3flGZm8d/Z3u/ut1rhFWZV12TOYtpfHLlz4HhKWH2T8YhyZCquqUJK/JLIycNnE5AOyDqnpRkn+X5MhVTjkmyfOTPLuqfrW7f2dvE2C9rXswAQ5490ry1KmLAGBN7pOVQ8nnk3w6yY1Jjs7s1q77z31/VJJ/V1UndPdL1r1K2APBBA48N2R26+Nq/GGWn7vzjKxuJ/CmVYy5PbM36T1glbUAsNxUv88/keT3k7yru68dfrn0HPBrMrvVfqeLqupD3b1llfXC6AQTOMAs3f742dWMrapbB11f7u5VzR1eNsnfZPaq752fDyc5Met/fzTAIWmC3+dXJnnJ3o466O4PV9Vjk7wjyY/NffXbVfWn3X3HgteFUQgmwCeTfFd37/K3brP3RABwEPjdJL+82kOhu/v2qvrpJJ9JcuxS95mZ7dpcsduJsI7GPvkdOMh0960rhRIADh7dff1qQ8ncnL9L8vZB9w+MVxUsRjABADh8fXTQPmWSKiCCCQDA4Wx4yOJqXzcMoxNMAAAOX/cdtLdPUgVEMAEAOJz9+KB95SRVQAQTAIDDUlX9WGZv4trpm0n+x0TlgGACAHC4qaqjk7x60P373b1jinogEUwAAA5Hr05y/7n2jUl+Y6JaIIlgAgBwWFk6WPEFg+6f7e6vTlEP7CSYAAAcJqrq0Ul+b9D9f3X326aoB+YJJgAAh4GqelCSdyY5aq77nUn+2TQVwXKCCQDAIa6qTk/y7iTHzXVfnuQnu/vbU9QEQ4IJAMAhrKpOTPKeJKfMdW9L8uTuvnWaqmBXggkAwCGqqo5L8t+S3G+u+5NJfqS7b56mKliZYAIAcAiqqk1J/jzJ98x1fy7J47v7K9NUBbsnmAAAHGKq6q5JtiR55Fz3l5M8rrtvmKYq2DPBBADgEFJVd0ny/yZ5/Fz315L8cHf/zTRVwd4JJgAAh4iqqiRvSPKUue5vJHlid181TVWwOoIJAMCh47eTPGeufWuSH+/uv5qoHlg1wQQA4BBQVf8qyS/Odd2e5Ond/T8mKgkWsmHqAoDpVdU9kxyzwlf3HrSPXjqkayU3dPftY9YFwOpU1T9N8huD7pcmuWoPv7dXcmt3f3msumARggmQzLb+f3oV4x6e5LrdfHdGks+OVRAAC3nmCn2vWPos4vIkj11zNbAP3MoFAABMTjABAAAmV909dQ0AAMBhzo4JAAAwOcEEAACYnGACAABMTjABAAAmJ5gAAACTE0wAAIDJCSYAAMDkBBMAAGByggkAADA5wQQAAJicYAIAAExOMAEAACYnmAAAAJMTTAAAgMkJJgAAwOQEEwAAYHKCCQAAMDnBBAAAmJxgAgAATE4wAQAAJieYAAAAkxNMAACAyf3/Z6gHfF27nScAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 259,
       "width": 403
      },
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(1, 1, tight_layout=True)\n",
    "\n",
    "labels = [\"T1\", \"T2\"]#, \"T3\"]\n",
    "\n",
    "x = np.arange(len(labels))  # the label locations\n",
    "\n",
    "width = 0.35  # the width of the bars\n",
    "\n",
    "r1 = ax.bar(x - width/2, y, width, label='Cluster')\n",
    "r2 = ax.bar(x + width/2, yy, width, label='!Cluster')\n",
    "ax.errorbar(x- width/2, y, yerr=yerr, fmt='.k');\n",
    "ax.errorbar(x+ width/2, yy, yerr=yyerr, fmt='.k');\n",
    "\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(labels)\n",
    "ax.set_ylim(0,1)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
